pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -203,6 +203,58 @@ type PersistentVolumeController struct {
 	// For testing only: hook to intercept CSI driver name <=> Intree plugin name mapping
 	// Not used when set to nil
 	csiNameFromIntreeNameHook func(pluginName string) (string, error)
+
+	// ThreadSafeStore caches start timestamp of operations
+	// (currently provision/prebinding/binding/deletion) for metric recording.
+	// Detailed lifecyle/key for each operation
+	// 1. provision metric
+	//     key:        claimKey + "-provision"
+	//     start time: user has NOT provide any volume ref in the claim AND
+	//                 there is no existing volume found for the claim,
+	//                 "provisionClaim" is called for a valid plugin/external provisioner
+	//                 to provision a volume
+	//     end time:   after a volume has been provisioned and bound to the claim successfully
+	//                 the corresponding timestamp entry will be deleted from cache
+	//     abort:      claim has not been bound to a volume yet but a claim deleted event
+	//                 has been received from API server
+	// 2. binding
```

Comment:
I really like the format of the comments, thanks!

I think we should combine binding and prebinding to the same operation.  I'm not sure how much value there is to break them down so much?  At a high level, I think it's good enough to have one measurement for provisioning, which invokes storage provider, and another measurement for binding, which only involves k8s controller.  wdyt @gnufied?

Reply:
I actually think it worth separating binding v.s. prebinding from user's perspective as they actually represents different semantics. For binding/provision, it applies to situations where users does not care which specific volume to use or pre-existence of a volume of the asked kind. For prebinding, users wants a specific volume to use. Though the code might not differ that much, the workflow does.

Reply:
There's 2 prebinding methods to consider:
1. PV.ClaimRef is set
2. PVC.volumeName is set

Reply:
for the first case, I'd rather still call it a binding? my understanding about prebinding is "I know which volume I wanted, and thus I want to pre-set it". Anyways its just naming, I am more than happy to group prebinding/binding as well.

Reply:
The first case is prebinding too. It's just coming from the other direction.  The admin has already reserved the PV for a specific PVC.

But it's actually more complicated than that.  When pv.claimref is set, it can be done either by scheduler, or by user.  If it's set by scheduler, then the BoundByController annotation is set.

Reply:
fair enough. then I will just combine binding with prebinding and call it "binding"

Reply:
no longer apply